fix(scim): Return the slug for the team name

AniketDas-Tekky opened the pull request:
We've deprecated use of the name field and use slugs exclusively for teams except for SCIM. This resulted in some customer escalations where they were confused where certain teams were coming from.

This fix makes it so that the name returned to the SCIM provider is the slug, matching the rest of the product.

Fixes ER-1519

Commit:
fix(scim): Return the slug for the team name

We've deprecated use of the name field and use slugs exclusively for teams except for SCIM. This resulted in some customer escalations where they were confused where certain teams were coming from.

This fix makes it so that the name returned to the SCIM provider is the slug, matching the rest of the product.

Commit:
Merge branch 'master' into adas/er-1519/return-slug-instead-of-name

Commit:
update tests

leedongwei reviewed (changes requested):
AFAIK the IdP does string-matching against the `displayName` key, so this change would break existing links if their name != slug.

Commit:
Fixes after discussion with Danny

Non-idp-provisioned teams will have their names reset to empty string as the field is deprecated.
IDP-provisioned teams will be allowed to keep their names as this is the team name on their systems. If there's a mismatch between name and slug, then the IDP acts as the source of truth and updates the slug to match the name
SCIM team details now returns different displayNames if they are IdP provisioned or not.
Finally, this locks team name changes from sentry if the team is idp provisioned.

wedamija commented on src/sentry/migrations/0429_resolve_mismatched_team_slugs_for_scim.py:
```diff
@@ -0,0 +1,45 @@
+
+    for team in RangeQuerySetWrapperWithProgressBar(Team.objects.all()):
+        if team.idp_provisioned and team.slug != slugify(team.name):
+            team.slug = slugify(team.name)
```
Is potentially changing team slugs going to break anything here?

AniketDas-Tekky replied:
Should be fine. It only affects teams provisioned through IdP recently. Looking at redash, I only see 6 instances where the team would be renamed.

Was there something else you're concerned about?

wedamija replied:
Nothing in particular. I'm just wary about renaming slugs since they tend to be used as primary keys externally. So these slugs aren't intended to match to existing value in idp or anything?

AniketDas-Tekky replied:
The slugs aren't, the Idp uses the id that we return. I can change the name as much as I want and it'll still recognize the team.

AniketDas-Tekky replied:
Actually, SCIM APIs don't even return a slug, they just return an ID and "displayName" which could be anything on the app side.

wedamija reviewed (commented):
Does the migration need to be run before the rest of the code is merged?

Commit:
fix tests and migration conflicts

Commit:
migration file

Commit:
rebase

Commit:
Merge branch 'master' into adas/er-1519/return-slug-instead-of-name

Commit:
migration conflict

Commit:
Removed the backfill of empty team name to non idp teams.

Turns out, that our current implementation of PATCH Scim Team solves the issue.

AniketDas-Tekky commented:
> > Does the migration need to be run before the rest of the code is merged?
> 
> Didn't get an answer here

It doesn't have to run before.

I actually just changed it now so that it won't change every team name. I just verified that our implementation of PATCH SCIM Team will fix older teams once an update is pushed.

AniketDas-Tekky committed:
Merge branch 'master' into adas/er-1519/return-slug-instead-of-name
